src/fine_list.ml

```diff
@@ -0,0 +1,148 @@
+type 'a node = { value : 'a; mutable next : 'a node option; lock : Mutex.t }
+type 'a t = { head : 'a node }
+
+let create_node value nextptr =
+  { value; next = nextptr; lock = Mutex.create () }
+
+let create key = { head = create_node key None }
+
+(* find the previous node <= key *)
+let find_previous_add t key =
+  let rec aux prev next =
+    match next with
+    | Some node when node.value > key -> prev
```

Comment:
Hmm... I wonder whether the logic and `add` and `remove` could be modified slightly so that only a single `find_previous` helper would be needed without impacting performance.